filmreel/src/response.rs

```
@@ -258,6 +256,73 @@ impl Validator {
         }
         Ok(())
     }
+
+    fn apply_unordered(
+        &self,
+        selector: Selector,
+        self_body: &mut Value,
+        other_body: &mut Value,
+    ) -> Result<(), FrError> {
+        let selection = selector(self_body).ok_or(FrError::ReadInstruction(
+            "selection missing from Frame body",
+        ))?;
+        match selection {
+            Value::Object(_) => Ok(()),
+            Value::Array(self_selection) => {
+                let other_selection = match selector(other_body) {
+                    Some(Value::Array(o)) => o,
+                    _ => return Ok(()),
+                };
+
+                // Create a lookup hashmap for self_selection.
+                let mut self_selection_hash: HashMap<Key, bool> = HashMap::new();
+                self_selection.iter().for_each(|s| {
+                    if let Ok(k) = to_key(&s) {
+                        self_selection_hash.insert(k, true);
+                    };
+                });
```

This pattern will swallow any errors returned by `to_key`.

We can instead call  `.collect()` on a map of results, unwrap it with `?` and call `.collect()` collect again on valid `Key` values _if_ we have no errors:

```rust
// Create a lookup hashmap for self_selection.
let self_selection_hash: HashMap<Key, bool> = self_selection
	.iter()
	.map(to_key) // Result<Key, Error>
	.collect::<Result<Vec<Key>, _>>()
	.map_err(|e| FrError::ReadInstructionf("apply_unordered", e.to_string()))?
	.into_iter() // IntoIterator<Item = Key>
	.map(|k| (k, true))
	.collect();
```

---

Whoa. I'm having Java8 flashbacks. 😰

---

@phernandez  🤷  There's also the go style of having the error be a separate result [using `partition()`](https://doc.rust-lang.org/rust-by-example/error/iter_result.html#collect-all-valid-values-and-failures-with-partition)

---

```rust
let (keys, errors): (Vec<_>, Vec<_>) = self_selection()
        .iter()
         ... // do the map
        .partition(Result::is_ok); // split based on if value is Ok(value)
errors.into_iter().collect()?; // unwrap errors here
```

---

Probably a cleaner more readable way of doing it as well 🤔